File: 백준 5639 TR 이진 검색 트리.cs
using System;
using System.Collections.Generic;

class Program {
    static List<int> data = new List<int>();

    static void PostOrder(int start, int end) {
        if (start > end) return;

        int div = end + 1;
        for (int i = start + 1; i <= end; i++) {
            if (data[start] < data[i]) {
                div = i;
                break;
            }
        }

        PostOrder(start + 1, div - 1);
        PostOrder(div, end);
        Console.WriteLine(data[start]);
    }

    static void Main() {
        string line;
        while ((line = Console.ReadLine()) != null && line != "") {
            data.Add(int.Parse(line));
        }

        PostOrder(0, data.Count - 1);
    }
}

File: 백준 5639 TR 이진 검색 트리.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;

public class Main {
    static ArrayList<Integer> data = new ArrayList<>();

    public static void postOrder(int start, int end) {
        if (start > end) return;

        int div = end + 1;
        for (int i = start + 1; i <= end; i++) {
            if (data.get(start) < data.get(i)) {
                div = i;
                break;
            }
        }

        postOrder(start + 1, div - 1);
        postOrder(div, end);
        System.out.println(data.get(start));
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        String line;
        while ((line = br.readLine()) != null && !line.isEmpty()) {
            data.add(Integer.parseInt(line));
        }

        postOrder(0, data.size() - 1);
    }
}
